File: applications/rtkiterativefdk/rtkiterativefdk.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkiterativefdk_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkIterativeFDKConeBeamReconstructionFilter.h"
#ifdef RTK_USE_CUDA
#  include "rtkCudaIterativeFDKConeBeamReconstructionFilter.h"
#endif

#include <itkStreamingImageFilter.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkiterativefdk, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
#endif

  // Projections reader
  using ReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto reader = ReaderType::New();
  rtk::SetProjectionsReaderFromGgo<ReaderType, args_info_rtkiterativefdk>(reader, args_info);

  if (args_info.verbose_flag)
    std::cout << "Reading... " << std::endl;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(reader->Update())

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Create reconstructed image
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkiterativefdk>(constantImageSource,
                                                                                         args_info);

  bool enforcePositivity = false;
  if (args_info.positivity_flag)
    enforcePositivity = true;
  else
    enforcePositivity = false;

  // Since the last template argument for IterativeFDKConeBeamReconstructionFilter is
  // double for the CPU version, and float for the CUDA one, we cannot have a single
  // pointer for both possibilities. In order to set the options only once,
  // we therefore create a macro, exactly as in rtkfdk.cxx
#define SET_IFDK_OPTIONS(f)                               \
  f->SetInput(0, constantImageSource->GetOutput());       \
  f->SetInput(1, reader->GetOutput());                    \
  f->SetGeometry(geometry);                               \
  SetForwardProjectionFromGgo(args_info, f.GetPointer()); \
  f->SetNumberOfIterations(args_info.niterations_arg);    \
  f->SetTruncationCorrection(args_info.pad_arg);          \
  f->SetHannCutFrequency(args_info.hann_arg);             \
  f->SetHannCutFrequencyY(args_info.hannY_arg);           \
  f->SetProjectionSubsetSize(args_info.subsetsize_arg);   \
  f->SetLambda(args_info.lambda_arg);                     \
  f->SetEnforcePositivity(enforcePositivity);             \
  f->SetDisableDisplacedDetectorFilter(args_info.nodisplaced_flag);

  // Create Iterative FDK filter and connect it
  using IFDKCPUType = rtk::IterativeFDKConeBeamReconstructionFilter<OutputImageType, OutputImageType, double>;
  IFDKCPUType::Pointer ifdk;
#ifdef RTK_USE_CUDA
  using IFDKCUDAType = rtk::CudaIterativeFDKConeBeamReconstructionFilter;
  IFDKCUDAType::Pointer ifdkCUDA;
#endif

  OutputImageType::Pointer IFDKOutputPointer;
  if (!strcmp(args_info.hardware_arg, "cpu"))
  {
    ifdk = IFDKCPUType::New();
    REPORT_ITERATIONS(ifdk, IFDKCPUType, OutputImageType)
    SET_IFDK_OPTIONS(ifdk);
    IFDKOutputPointer = ifdk->GetOutput();
  }
#ifdef RTK_USE_CUDA
  else if (!strcmp(args_info.hardware_arg, "cuda"))
  {
    ifdkCUDA = IFDKCUDAType::New();
    REPORT_ITERATIONS(ifdkCUDA, IFDKCPUType, OutputImageType)
    SET_IFDK_OPTIONS(ifdkCUDA);
    IFDKOutputPointer = ifdkCUDA->GetOutput();
  }
#endif

  // Write
  if (args_info.verbose_flag)
    std::cout << "Reconstructing and writing... " << std::endl;

  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(IFDKOutputPointer, args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkiterativefdk/rtkiterativefdk.py
#!/usr/bin/env python
import argparse
import sys
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description=(
            "Reconstructs a 3D volume from a sequence of projections "
            "[Feldkamp, David, Kress, 1984]."
        )
    )

    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--hardware",
        help="Hardware used for computation",
        choices=["cpu", "cuda"],
        default="cpu",
    )
    parser.add_argument(
        "--subsetsize",
        help="Streaming option: number of projections processed at a time",
        type=int,
        default=16,
    )
    parser.add_argument(
        "--niterations", "-n", type=int, default=3, help="Number of iterations"
    )
    parser.add_argument(
        "--lambda",
        "-l",
        dest="lambda_",
        type=float,
        default=0.3,
        help="Convergence factor",
    )
    parser.add_argument(
        "--positivity",
        help="Enforces positivity during the reconstruction",
        action="store_true",
    )
    parser.add_argument(
        "--nodisplaced",
        help="Disable the displaced detector filter",
        action="store_true",
    )

    # Ramp filter options
    ramp_group = parser.add_argument_group("Ramp filter")
    ramp_group.add_argument(
        "--pad",
        help="Data padding parameter to correct for truncation",
        type=float,
        default=0.0,
    )
    ramp_group.add_argument(
        "--hann",
        help="Cut frequency for hann window in ]0,1] (0.0 disables it)",
        type=float,
        default=0.0,
    )
    ramp_group.add_argument(
        "--hannY",
        help="Cut frequency for hann window in ]0,1] (0.0 disables it)",
        type=float,
        default=0.0,
    )

    rtk.add_rtkprojectors_group(parser)
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtk3Doutputimage_group(parser)
    rtk.add_rtkiterations_group(parser)

    return parser


def process(args_info: argparse.Namespace):
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Check on hardware parameter
    if not hasattr(itk, "CudaImage") and args_info.hardware == "cuda":
        print("The program has not been compiled with CUDA option.")
        sys.exit(1)

    # Projections reader
    reader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)

    if args_info.verbose:
        print("Reading... ")
    reader.Update()

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry information from {args_info.geometry}...")
    geometry = rtk.read_geometry(args_info.geometry)

    # Create reconstructed image
    constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
    rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)

    enforcePositivity = bool(args_info.positivity)

    # Create Iterative FDK filter and connect it
    if args_info.hardware == "cuda":
        ifdk = rtk.CudaIterativeFDKConeBeamReconstructionFilter.New()
        ifdk.SetInput(0, itk.cuda_image_from_image(constantImageSource.GetOutput()))
        ifdk.SetInput(1, itk.cuda_image_from_image(reader.GetOutput()))
    else:
        IFDKCPUType = rtk.IterativeFDKConeBeamReconstructionFilter[
            OutputImageType, OutputImageType, itk.D
        ]
        ifdk = IFDKCPUType.New()
        ifdk.SetInput(0, constantImageSource.GetOutput())
        ifdk.SetInput(1, reader.GetOutput())

    # Common options
    ifdk.SetGeometry(geometry)
    rtk.SetForwardProjectionFromArgParse(args_info, ifdk)
    ifdk.SetNumberOfIterations(args_info.niterations)
    ifdk.SetTruncationCorrection(args_info.pad)
    ifdk.SetHannCutFrequency(args_info.hann)
    ifdk.SetHannCutFrequencyY(args_info.hannY)
    ifdk.SetProjectionSubsetSize(args_info.subsetsize)
    ifdk.SetLambda(args_info.lambda_)
    ifdk.SetEnforcePositivity(enforcePositivity)
    ifdk.SetDisableDisplacedDetectorFilter(args_info.nodisplaced)

    # Iterations reporting
    rtk.SetIterationsReportFromArgParse(args_info, ifdk)

    # Run the filter and write
    if args_info.verbose:
        print("Reconstructing and writing... ")

    # Write
    writer = itk.ImageFileWriter[OutputImageType].New()
    writer.SetFileName(args_info.output)
    writer.SetInput(ifdk.GetOutput())
    writer.Update()


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
